packages/graphql-live-query-patch/src/LiveExecutionPatch.ts

```diff
@@ -0,0 +1,11 @@
+import { Operation } from "fast-json-patch";
+import { ExecutionResult } from "graphql";
+
+export type LiveExecutionPatch = {
+  revision?: number;
+  data?: ExecutionResult["data"];
+  errors?: ExecutionResult["errors"];
+  extensions?: ExecutionResult["extensions"];
+  patch?: Operation[];
+  isLivePatch?: true;
```

Comment:
I think I agree with the addition of `isLivePatch` but only after a heavy dose of uncertainty. So just in case anyone else comes along with a similar dilemma here were my pros and cons before I settled on the pros side.

On the pros side:

- `operation.isLivePatch == true` is simple in both Rust, JS and.. well.. anywhere I can think of.
- it's already implemented this way and I really prefer things that are already built ;)

On the cons side:

- it adds a little extra overhead to each message
- `operation.patch != null` is equivalent in JS for well-formed messages
- to robustly check for malformed messages an `operiation.isLivePatch && operation.patch != null` check is probably still necessary in JS even if isLivePatch is true
- in Rust we could model this as an enum (Algebraic Data Type) just using the data and patch fields' existence / lack there of so there's no need fo isLivePatch over there.

Reply:
The existence of the patch field could definitely be the alternative for isLivePatch!

Reply:
Btw this is the signature of the patch result used for defer/stream:

```ts
/**
 * The result of an asynchronous GraphQL patch.
 *
 *   - `errors` is included when any errors occurred as a non-empty array.
 *   - `data` is the result of the additional asynchronous data.
 *   - `path` is the location of data.
 *   - `hasNext` is true if a future payload is expected.
 *   - `label` is the label provided to @defer or @stream.
 *   - `extensions` is reserved for adding non-standard properties.
 */
export interface ExecutionPatchResult<
  TData = { [key: string]: any },
  TExtensions = { [key: string]: any }
> {
  errors?: ReadonlyArray<GraphQLError>;
  data?: TData | null;
  path?: ReadonlyArray<string | number>;
  label?: string;
  hasNext: boolean;
  extensions?: TExtensions;
}
```

Reply:
See https://github.com/n1ru4l/graphql-live-query/pull/243/commits/dc2f473ea0365520459a5569248be7fc70ee913e

Reply:
The only reason I see for keeping that field is so that the client knows whether the thing he gets is a patch that can be applied or sth else.

Reply:
> Btw this is the signature of the patch result used for defer/stream:
> 
> ```ts
> /**
>  * The result of an asynchronous GraphQL patch.
>  *
>  *   - `errors` is included when any errors occurred as a non-empty array.
>  *   - `data` is the result of the additional asynchronous data.
>  *   - `path` is the location of data.
>  *   - `hasNext` is true if a future payload is expected.
>  *   - `label` is the label provided to @defer or @stream.
>  *   - `extensions` is reserved for adding non-standard properties.
>  */
> export interface ExecutionPatchResult<
>   TData = { [key: string]: any },
>   TExtensions = { [key: string]: any }
> > {
>   errors?: ReadonlyArray<GraphQLError>;
>   data?: TData | null;
>   path?: ReadonlyArray<string | number>;
>   label?: string;
>   hasNext: boolean;
>   extensions?: TExtensions;
> }
> ```

Oh, interesting. Why this format vs. RFC 6902 JSON Patch? (This is instead of JSON Patch right?)

Reply:
This format is only meant for deferring the load of a fragment and incrementally loading a list. There is no intention in being flexible and keeping the state in sync.

```gql
query user($id: ID!) {
  user(id: $id) {
    id
    login
    ... on User {
      posts(first: 4) @defer(label: "userPosts") {
        id
        title
      }
    }
  }
}
```

Would result in the following responses:

```json
{
  "data": {
    "user": {
      "id": 1,
      "login": "Pete",
      "posts": null
    }
  },
  "hasNext": true
}
```

```json
{
  "data": { "posts": [{ "id": 1, "title": "1" }, { "id": 2, "title": "2" }] },
  "path": ["user"],
  "hasNext": false
}
```

Reply:
oh whoops nope, I just misread what you were saying. I was thinking that was the format for `@live` patches for some reason.

Reply:
Yeah, disregard that. Your doing exactly what I was looking for right here: https://github.com/n1ru4l/graphql-live-query/pull/243/files#diff-c2511c5651d368d9d759f75f525bd864c336d2110468f50b64c10218f9bcdcaeR64

Reply:
:+1:

Reply:
BTW defer and stream patches can be applied by relay automatically (https://github.com/facebook/relay/blob/master/packages/relay-runtime/store/RelayModernQueryExecutor.js)

For the JSON patches we need to apply the patch before forwarding the execution result to the store. 

With a custom QueryExecutor the JSON patches could be applied to the store directly. However, then you would have new issues like, what if two active operations modify the same resources in the cache. A operation that would add an item to an array could result in the item being added twice etc.

Reply:
what if the live query store would be smarter on server side an keep a normalized result of all the resources the operations of the client selects and then send only patches for such resources to the client that have change. Sounds a bit complicated though 🤔

Reply:
BTW @D1plo1d have you seen https://github.com/apollographql/real-time-federation
I think there are also a few nice concepts.

A drawback of this live query implementation, in general, is that live queries that select the same resource will all receive an update patch. Although only one patch might be sufficient (depending on the selection set) as the client cache keeps the consumed queries synced with the cache 🤔. A stream that publishes information about ALL selected resources could be more efficient and have less overhead, but would probably result in having a normalized state of the client on the server which is used for determining the patch operations.

Reply:
Yeah, the question of how different operations might modify the state of the same object is interesting. Also real-time-federation is quite a concept - hadn't seen that one before.

So I think I am leaning towards denormalized query tree patches instead of sending normalized patches over the wire:

1. Because I suspect we may be optimizing a cold path that results in negligible overhead - AFAIK my queries most commonly reference each object exactly once. This is entirely based on my usage and my terrible memory YMMV.
2. Keeping a minimal `@live` graphql client simpler is important IMO to achieving support across implementations & in multiple languages.

I'm also leaning towards JSON patches before the store for a couple reasons:

1. If we go with denormalized query tree patches then conflicting updates will at least not result in the double array addition scenario you outlined in the custom QueryExecutor example.
2. At worst an outdated result will be applied after a newer result and the cache will be out of date. We already have a similar problem if two non-live but asynchronous query results come back in reverse order. It's weird but at least we're being consistently weird?
3. Unlike a custom QueryExecutor we can do this in an apollo link without forking apollo client itself

Reply:
Yes after thinking about it further and your thoughts I also agree that that idea might get too complicated 😅

Reply:
Also real time federation actually just feels like a hack for missing subscription support inside federation the more I look at it 😅

Reply:
But I like the directive stuff which could be used for an abstraction on top of the in-memory-live-query-store package